### ISSUE TITLE
Brokered AddressSpace: creation of sender link takes multiple seconds

### ISSUE THREAD

**Issue by calohmn** [EnMasseProject/enmasse#1430]
We've done some load tests using a brokered address space, sending 500 byte messages from 300 clients, each creating an AMQP connection.

The first requests take multiple seconds. This is due to the sender link creation taking up to 12s.

(EnMasse 0.20.0 with Artemis 2.6.1 image using OpenSSL as sslProvider.)

**Comment by calohmn** [comment]
Screenshot of a JVisualVM CPU sampling result:
Thread-4 (activemq-netty-threads), Total Time: 67.373 ms (100%), Total Time (CPU): 67.266 ms (100%):
![artemis_thread4activemq-netty-threads_sasldelegatinglogin](https://user-images.githubusercontent.com/891931/42503963-71083652-843a-11e8-9b5a-54656af3c8ca.png)

Log output with method invocations of `SaslDelegatingLogin` from a local test with *one* sender link creation:
````
10:26:16,636 [SaslDelegatingLogin] login
10:26:16,647 [SaslDelegatingLogin] -writeToNetwork
10:26:16,647 [SaslDelegatingLogin] --writing 8 bytes
10:26:16,722 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:16,723 [SaslDelegatingLogin] chooseSaslMechanismAndSendInit
10:26:16,723 [SaslDelegatingLogin] -readFromNetwork
10:26:16,730 [SaslDelegatingLogin] --read 80 bytes
10:26:16,731 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:16,731 [SaslDelegatingLogin] -writeToNetwork
10:26:16,732 [SaslDelegatingLogin] --writing 53 bytes
10:26:16,732 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:16,732 [SaslDelegatingLogin] chooseSaslMechanismAndSendInit DONE
10:26:16,732 [SaslDelegatingLogin] performSaslSteps
10:26:16,732 [SaslDelegatingLogin] -readFromNetwork
10:26:16,812 [SaslDelegatingLogin] --read 24 bytes
10:26:16,813 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:16,813 [SaslDelegatingLogin] performSaslSteps DONE
10:26:16,813 [SaslDelegatingLogin] performConnectionOpen
10:26:16,814 [SaslDelegatingLogin] -writeToNetwork
10:26:16,814 [SaslDelegatingLogin] --writing 70 bytes
10:26:16,814 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:16,814 [SaslDelegatingLogin] -readFromNetwork
10:26:16,818 [SaslDelegatingLogin] --read 332 bytes
10:26:16,820 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:16,820 [SaslDelegatingLogin] performConnectionOpen DONE
10:26:16,820 [SaslDelegatingLogin] login DONE
10:26:16,837 [SaslDelegatingLogin] login
10:26:16,842 [SaslDelegatingLogin] -writeToNetwork
10:26:16,842 [SaslDelegatingLogin] --writing 8 bytes
10:26:16,920 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:16,921 [SaslDelegatingLogin] chooseSaslMechanismAndSendInit
10:26:16,921 [SaslDelegatingLogin] -readFromNetwork
10:26:16,930 [SaslDelegatingLogin] --read 80 bytes
10:26:16,930 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:16,931 [SaslDelegatingLogin] -writeToNetwork
10:26:16,932 [SaslDelegatingLogin] --writing 53 bytes
10:26:16,933 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:16,933 [SaslDelegatingLogin] chooseSaslMechanismAndSendInit DONE
10:26:16,933 [SaslDelegatingLogin] performSaslSteps
10:26:16,933 [SaslDelegatingLogin] -readFromNetwork
10:26:17,035 [SaslDelegatingLogin] --read 24 bytes
10:26:17,036 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:17,036 [SaslDelegatingLogin] performSaslSteps DONE
10:26:17,036 [SaslDelegatingLogin] performConnectionOpen
10:26:17,036 [SaslDelegatingLogin] -writeToNetwork
10:26:17,037 [SaslDelegatingLogin] --writing 70 bytes
10:26:17,039 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:17,039 [SaslDelegatingLogin] -readFromNetwork
10:26:17,041 [SaslDelegatingLogin] --read 332 bytes
10:26:17,044 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:17,044 [SaslDelegatingLogin] performConnectionOpen DONE
10:26:17,045 [SaslDelegatingLogin] login DONE
10:26:17,047 [SaslDelegatingLogin] login
10:26:17,048 [SaslDelegatingLogin] -writeToNetwork
10:26:17,049 [SaslDelegatingLogin] --writing 8 bytes
10:26:17,106 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:17,108 [SaslDelegatingLogin] chooseSaslMechanismAndSendInit
10:26:17,109 [SaslDelegatingLogin] -readFromNetwork
10:26:17,111 [SaslDelegatingLogin] --read 80 bytes
10:26:17,112 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:17,113 [SaslDelegatingLogin] -writeToNetwork
10:26:17,119 [SaslDelegatingLogin] --writing 53 bytes
10:26:17,127 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:17,132 [SaslDelegatingLogin] chooseSaslMechanismAndSendInit DONE
10:26:17,133 [SaslDelegatingLogin] performSaslSteps
10:26:17,133 [SaslDelegatingLogin] -readFromNetwork
10:26:17,187 [SaslDelegatingLogin] --read 24 bytes
10:26:17,188 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:17,189 [SaslDelegatingLogin] performSaslSteps DONE
10:26:17,189 [SaslDelegatingLogin] performConnectionOpen
10:26:17,194 [SaslDelegatingLogin] -writeToNetwork
10:26:17,194 [SaslDelegatingLogin] --writing 70 bytes
10:26:17,194 [SaslDelegatingLogin] -writeToNetwork DONE
10:26:17,195 [SaslDelegatingLogin] -readFromNetwork
10:26:17,196 [SaslDelegatingLogin] --read 332 bytes
10:26:17,197 [SaslDelegatingLogin] -readFromNetwork DONE
10:26:17,197 [SaslDelegatingLogin] performConnectionOpen DONE
10:26:17,201 [SaslDelegatingLogin] login DONE
````

**Comment by rgodfrey** [comment]
> I wonder why `login()` gets called 3 times (?).

@calohmn  IIRC Artemis calls login at the initial time of connection creation, and then again once each time a session is opened.  This is obviously a bit pointless (and likely originates from how other protocols than AMQP do their authentication), but I would hope it would not normally be a bottleneck (since in general sessions are not normally created and destroyed frequently).  If this is proving to be an issue for your use case there's maybe some trickery I can do to stop Artemis trying to re-authenticate for an already authenticated connection.

**Comment by calohmn** [comment]
@rgodfrey In general, usage of the brokered address space is not planned for now in our case. It was just part of a comparison with the standard address space concerning performance figures. And there, the difference in times for the link creation was conspicuous. With the same test setup, link creation with qpid as target took 260ms at most. So this issue is for reporting that finding.

**Comment by rgodfrey** [comment]
@calohmn yeah - looking at the Artemis code it seems it may well try to re-authenticate on every link not just every session *sigh*.  I have a patch which I'll put up shortly that attempts to use a horrible hack to get around this issue; however really the issue should be solved in Artemis.

**Comment by k-wall** [comment]
Believed to be fixed by #1432 - closing.
